Plugins/EDM4hep/src/EDM4hepUtil.cpp

```diff
@@ -133,36 +129,31 @@ Parameters convertTrackParametersToEdm4hep(const Acts::GeometryContext& gctx,
     // We need to convert to the target parameters
     // Keep the free parameters around we might need them for the covariance
     // conversion
-    freePars = makeFreePars();
-    targetPars = Acts::detail::transformFreeToBoundParameters(freePars.value(),
+    Acts::FreeVector freePars = Acts::detail::transformBoundToFreeParameters(
+        params.referenceSurface(), gctx, params.parameters());
+    targetPars = Acts::detail::transformFreeToBoundParameters(freePars,
                                                               *refSurface, gctx)
                      .value();
+
+    if (targetCov) {
+      // We need to convert the covariance as well
+      Acts::BoundToFreeMatrix boundToFree =
+          params.referenceSurface().boundToFreeJacobian(gctx, targetPars);
+      Acts::FreeToBoundMatrix freeToBound =
+          refSurface->freeToBoundJacobian(gctx, freePars);
+      Acts::BoundSquareMatrix boundToBound = freeToBound * boundToFree;
+      targetCov = boundToBound * targetCov.value() * boundToBound.transpose();
+    }
   }
 
   Parameters result;
   result.surface = refSurface;
 
   // Only run covariance conversion if we have a covariance input
-  if (params.covariance()) {
-    auto boundToFree =
-        refSurface->boundToFreeJacobian(gctx, params.parameters());
-    Acts::FreeMatrix freeCov =
-        boundToFree * params.covariance().value() * boundToFree.transpose();
-
-    // ensure we have free pars
-    if (!freePars.has_value()) {
-      freePars = makeFreePars();
-    }
-
-    Acts::CovarianceCache covCache{freePars.value(), freeCov};
-    auto [varNewCov, varNewJac] = Acts::transportCovarianceToBound(
-        gctx, *refSurface, freePars.value(), covCache);
-    auto targetCov = std::get<Acts::BoundSquareMatrix>(varNewCov);
-
+  if (targetCov) {
     Acts::ActsSquareMatrix<6> J = jacobianToEdm4hep(
         targetPars[eBoundTheta], targetPars[eBoundQOverP], Bz);
-    Acts::ActsSquareMatrix<6> cIn = targetCov.template topLeftCorner<6, 6>();
-    result.covariance = J * cIn * J.transpose();
+    result.covariance = J * targetCov.value() * J.transpose();
```

Comment:
@paulgessinger I tried to implement what we discussed. not sure if this is correct tho :smile: